recc 1.2.21

sakeeb opened the pull request:
RECC is a compiler launcher that caches the results on compilation and link command, and optionally forward them to a remote execution service.

<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+    File.write(config_directory / "recc.conf", ERB.new(recc_conf_tmpl).result(binding))
+  end
+
+  def generate_recc_server(bin_directory)
```
This can probably be inlined into the `install` method.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+  end
+
+  def generate_recc_server(bin_directory)
+    (bin_directory / "recc-server").write <<~EOS
```
```suggestion
    (bin_directory/"recc-server").write <<~EOS
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+  end
+
+  def generate_recc_wrapper(bin_directory, compiler)
+    (bin_directory / "recc-#{compiler}").write <<~EOS
```
```suggestion
    (bin/"recc-#{compiler}").write <<~EOS
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+    EOS
+  end
+
+  def generate_recc_wrapper(bin_directory, compiler)
```
```suggestion
  def generate_recc_wrapper(compiler)
```
This should also be in-lined in the `install` method.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+    EOS
+  end
+
+  def buildbox_cmake_args
+    %W[
+      -DCASD=ON
+      -DCASD_BUILD_BENCHMARK=OFF
+      -DCASDOWNLOAD=OFF
+      -DCASUPLOAD=OFF
+      -DFUSE=OFF
+      -DLOGSTREAMRECEIVER=OFF
+      -DLOGSTREAMTAIL=OFF
+      -DOUTPUTSTREAMER=OFF
+      -DRECC=ON
+      -DREXPLORER=OFF
+      -DRUMBA=OFF
+      -DRUN_BUBBLEWRAP=OFF
+      -DRUN_HOSTTOOLS=ON
+      -DRUN_OCI=OFF
+      -DRUN_USERCHROOT=OFF
+      -DTREXE=OFF
+      -DWORKER=OFF
+      -DRECC_CONFIG_PREFIX_DIR=#{etc}/recc
+    ]
+  end
+
+  def install
+    system "cmake", "-S", ".", "-B", "build", *buildbox_cmake_args, *std_cmake_args
```
```suggestion
  def install
    buildbox_cmake_args = %W[
      -DCASD=ON
      -DCASD_BUILD_BENCHMARK=OFF
      -DCASDOWNLOAD=OFF
      -DCASUPLOAD=OFF
      -DFUSE=OFF
      -DLOGSTREAMRECEIVER=OFF
      -DLOGSTREAMTAIL=OFF
      -DOUTPUTSTREAMER=OFF
      -DRECC=ON
      -DREXPLORER=OFF
      -DRUMBA=OFF
      -DRUN_BUBBLEWRAP=OFF
      -DRUN_HOSTTOOLS=ON
      -DRUN_OCI=OFF
      -DRUN_USERCHROOT=OFF
      -DTREXE=OFF
      -DWORKER=OFF
      -DRECC_CONFIG_PREFIX_DIR=#{etc}/recc
    ]

    system "cmake", "-S", ".", "-B", "build", *buildbox_cmake_args, *std_cmake_args
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+
+    generate_recc_server bin
+
+    generate_recc_conf prefix / "etc/recc/"
```
```suggestion
    generate_recc_conf prefix/"etc/recc/"
```
Note that writing this to `prefix/"etc"` means that users will not be able to write their own config files to `etc/"recc"`.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+            -c "add ProgramArguments: string 512" \\
+            -c "add ProgramArguments: string unlimited" \\
+            -c "add RunAtLoad bool true" && \\
+          sudo chown root:wheel /Library/LaunchDaemons/limit.maxfiles.plist && \\
+          sudo chmod 644 /Library/LaunchDaemons/limit.maxfiles.plist && \\
+          sudo launchctl load -w /Library/LaunchDaemons/limit.maxfiles.plist#{" "}
+        EOS
+      else
+        launchctl_help = ""
+      end
+      (general_info + launchctl_help)
+    end
+  end
+
+  def caveats
+    <<~EOS
+      The recc configuration is read in the following order.
+        1. Environment variables (see: https://buildgrid.gitlab.io/recc/configuration-variables.html)
+        2. $(pwd)/recc/recc.conf
+        3. {$RECC_CONFIG_DIRECTORY}/recc.conf
+        4. ~/.recc/recc.conf
+        5. #{etc}/recc/recc.conf
+        6. #{bin}/../etc/recc/recc.conf
+      If an option is set in multiple places, the one higher up in the list will be effective.
+
+      To launch a compiler with recc, set the following variables:
+        CC=#{bin}/recc-cc
+        CXX=#{bin}/recc-c++
+
+      For CMake builds, an alternative option is to use CMAKE_<LANG>_COMPILER/LINKER_LAUNCHER.
+      Please do not set both CC/CXX and CMAKE_<LANG>COMPILER?LINKER_LAUNCHER variables to recc.
+        CMAKE_C_COMPILER_LAUNCHER=#{bin}/recc
+        CMAKE_CXX_COMPILER_LAUNCHER=#{bin}/recc
+        CMAKE_C_LINKER_LAUNCHER=#{bin}/recc
+        CMAKE_CXX_LINKER_LAUNCHER=#{bin}/recc
+
+      Start the recc service to use the out-of-the-box configuration (see instructions below).
+      #{macos_caveats}
+    EOS
+  end
```
```suggestion
```
These caveats don't seem to be Homebrew-specific, so let's leave them out please, thanks.

sakeeb replied:
Make sense. Remove the non home-brew specific portion and only kept the portion about update ulimit when running the service.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+  end
+
+  test do
+    system bin/"recc", "--version"
```
Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

sakeeb replied:
Updated to test the feature.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+  end
+
+  service do
+    run [opt_bin / "recc-server"]
+    keep_alive true
+    working_dir var / "recc"
+    log_path var / "log/casd.log"
+    error_log_path var / "log/casd-error.log"
```
```suggestion
    run [opt_bin/"recc-server"]
    keep_alive true
    working_dir var/"recc"
    log_path var/"log/casd.log"
    error_log_path var/"log/casd-error.log"
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+    system bin/"recc", "--version"
+  end
+
+  service do
```
Please this before the `test do` block, please.

sakeeb replied:
Moved it up

carlocab reviewed (commented):
Thanks for the PR. A few suggestions.

sakeeb commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+      instance=recc-server
+
+      # Identifier to correlate all recc invocations#{" "}
+      correlated_invocations_id=<%= #{`hostname`.strip} %>
```
Thank you for pointing out. I was trying get an identifier unique to the host where the command will be run. I removed this option now.

sakeeb commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+    EOS
+
+    config_directory.mkpath
+    File.write(config_directory / "recc.conf", ERB.new(recc_conf_tmpl).result(binding))
```
Removed ERB. Thank you for the detailed review.

sakeeb commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,231 @@
+    end
+  end
+
+  def generate_recc_conf(config_directory)
```
`generate_recc_conf`  is called from `install` method. Are you suggesting me to move the whole code block inside the `install` method instead of wrapping it in another method and calling from there ?

carlocab replied:
Yes, exactly.

sakeeb replied:
Done.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,188 @@
+
+    generate_recc_conf prefix/"etc/recc/"
+  end
+
+  def caveats
+    on_macos do
+      <<~EOS
+        If you are starting the `recc` service via `brew service start`, please ensure the maxfiles limit
+        for `launchctl` is larger than 364. If you are running `recc-server` directly, please ensure
+        `ulimit -Sn` is larger than 364.
+
+         To increase the limit and make the change persist across reboots, you can create a launch daemon
+         using the following steps:
+
+         sudo /usr/libexec/PlistBuddy /Library/LaunchDaemons/limit.maxfiles.plist \\
+          -c "add Label string com.launchd.maxfiles" \\
+          -c "add ProgramArguments array" \\
+          -c "add ProgramArguments: string launchctl" \\
+          -c "add ProgramArguments: string limit" \\
+          -c "add ProgramArguments: string maxfiles" \\
+          -c "add ProgramArguments: string 512" \\
+          -c "add ProgramArguments: string unlimited" \\
+          -c "add RunAtLoad bool true" && \\
+         sudo chown root:wheel /Library/LaunchDaemons/limit.maxfiles.plist && \\
+         sudo chmod 644 /Library/LaunchDaemons/limit.maxfiles.plist && \\
+         sudo launchctl load -w /Library/LaunchDaemons/limit.maxfiles.plist
+      EOS
+    end
+  end
```
```suggestion
```
This isn't particular to Homebrew; please remove it.

sakeeb replied:
Removed.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+    ]
+
+    system "cmake", "-S", ".", "-B", "build", *buildbox_cmake_args, *std_cmake_args
+    system "cmake", "--build", "build", "-j", ENV.make_jobs.to_s
```
```suggestion
    system "cmake", "--build", "build"
```
You don't need this, `brew` sets `-j` appropriately in `MAKEFLAGS`.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+    system "cmake", "--build", "build", "-j", ENV.make_jobs.to_s
+    system "cmake", "--install", "build"
+
+    generate_recc_wrapper = lambda do |compiler|
+      (bin/"recc-#{compiler}").write <<~EOS
+        #!/usr/bin/env sh
+
+        #{bin}/recc $(command -v #{compiler}) "$@"
+      EOS
+    end
+
+    generate_recc_wrapper.call("cc")
+    generate_recc_wrapper.call("c++")
+    generate_recc_wrapper.call("gcc")
+    generate_recc_wrapper.call("g++")
+    generate_recc_wrapper.call("clang")
+    generate_recc_wrapper.call("clang++")
```
```suggestion
    %w[cc c++ gcc g++ clang clang++].each do |compiler|
      (bin/"recc-#{compiler}").write <<~EOS
        #!/usr/bin/env sh

        #{bin}/recc $(command -v #{compiler}) "$@"
      EOS
    end
```
You don't need a `lambda` here. Why aren't these wrappers created by the `install` target?

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+    generate_recc_wrapper.call("clang++")
+
+    # Generate recc-server start helper script
+    (bin/"recc-server").write <<~EOS
```
Why isn't this created by the `install` target?

sakeeb replied:
Although it's named recc-server here, the buildbox-casd binary is not specific to recc; rather, it is a general-purpose tool. As intended by the upstream project, recc-server serves as a wrapper script tailored to fit the recc use case.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+      # 2. ${cwd}/recc/recc.conf
+      # 3. {$RECC_CONFIG_DIRECTORY}/recc.conf
+      # 4. ~/.recc/recc.conf
+      # 5. #{etc}/recc/recc.conf
+      # 6. #{prefix}/etc/recc/recc.conf
```
This doesn't really make sense, since `etc/"recc/recc.conf"` will be a symlink to `prefix/"etc/recc/recc.conf"`.

sakeeb replied:
Thank you for pointing that out. I removed option 6

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+    # Generate recc.conf
+    config_directory = prefix/"etc/recc/"
+    config_directory.mkpath
+    (config_directory/"recc.conf").write <<~EOS
```
Is there no default `recc.conf` supplied in the upstream repo?

sakeeb replied:
There is no default recc.conf in upstream, as it can be used with any [REAPI](https://github.com/bazelbuild/remote-apis?tab=readme-ov-file#servers) server. It's mainly used in conjunction with a REAPI server attached to a build server. This formula aims to provide an out-of-the-box working experience, as configuring a remote server can be a daunting task for many developers

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+      retry_limit=0
+
+      # Platform properties
+      REMOTE_PLATFORM_OS=<%= platform_os %>
```
```suggestion
      REMOTE_PLATFORM_OS=#{OS.mac? ? "macos" : OS.kernel_name.downcase }
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+
+      # Platform properties
+      REMOTE_PLATFORM_OS=<%= platform_os %>
+      REMOTE_PLATFORM_ISA=<%= Hardware::CPU.arch.to_s %>
```
```suggestion
      REMOTE_PLATFORM_ISA=#{Hardware::CPU.arch}
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+    EOS
+
+    # Wait for the server to start
+    sleep 2 unless File.exist?(recc_cache_dir/"casd.sock")
```
```suggestion
    sleep 2 unless (recc_cache_dir/"casd.sock").exist?
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+    sleep 2 unless File.exist?(recc_cache_dir/"casd.sock")
+
+    # Override default values of server and log_level
+    recc_conf="RECC_SERVER=unix://#{recc_cache_dir}/casd.sock RECC_LOG_LEVEL=info"
+    recc_test="#{recc_conf} #{bin}/recc-cc -c #{test_file} 2>&1"
```
```suggestion
    ENV["RECC_SERVER"] = "unix://#{recc_cache_dir}/casd.sock"
    ENV["RECC_LOG_LEVEL"] = "info"
    recc_test="#{bin}/recc-cc -c #{test_file} 2>&1"
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,196 @@
+
+  on_linux do
+    depends_on "pkg-config" => :build
+    depends_on "util-linux" => :build
```
```suggestion
    depends_on "util-linux"
```
Looks like you need this at runtime.

sakeeb commented:
@carlocab, thank you for your detailed review. It really helped to trim down a lot of unnecessary content. I have addressed all the review comments. What would be the next step?

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,181 @@
+      # Get current soft limit for open files
+      current_limit=$(ulimit -Sn)
+
+      REQUIRED_FILES_LIMIT=384
```
Where does `384` come from?

sakeeb replied:
The `buildbox-casd` process will need 364 or more file descriptors. As recc-client connects to it, the number of file descriptors goes up. I used 20 as a reasonable upper bound

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,181 @@
+      REQUIRED_FILES_LIMIT=384
+      # Check if the current limit is less than REQUIRED_FILES_LIMIT
+      if [ "$current_limit" -lt ${REQUIRED_FILES_LIMIT} ]; then
+        ulimit -Sn ${REQUIRED_FILES_LIMIT}
```
It doesn't seem right for us to write a script that mutates the user's state when run.

If this is required, then users should know to do this from reading upstream documentation. It _might_ be okay to set it in the `service` block by setting the `SoftResourceLimits` property of the generated plist. (Though this may not yet be implemented in `brew`.)

sakeeb replied:
The updated ulimit only applies to the `recc-server`. I confirmed it by running the following

```bash
ulimit -Sn; ulimit -Sn 256; ulimit -Sn;recc-server &; pid=$! ; ulimit -Sn; kill $pid;ulimit -Sn
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,181 @@
+        # 1. Environment variables
+        # 2. ${cwd}/recc/recc.conf
+        # 3. {$RECC_CONFIG_DIRECTORY}/recc.conf
+        # 4. ~/.recc/recc.conf
+        # 5. #{etc}/recc.conf
+        ##
+
+        # Protocol version used for communicating with the REAPI server
+        reapi_version=2.2
+
+        # The URI of the server
+        server=unix://#{var}/recc/casd/casd.sock
+        # Instance name provided by the server
+        instance=recc-server
+
+        # Use cache-only mode, build anything not available in the cache locally
+        cache_only=1
+        # Upload results from the local builds to cache
+        cache_upload_local_build=1
+        # Upload failed results from the local builds to cache
+        cache_upload_failed_build=1
+
+        # Enable for link commands
+        link=1
+        # Use cache-only mode for link commands
+        link_cache_only=1
+
+        # Report all entries returned by the dependency command, even if they are absolute paths
+        deps_global_paths=1
+        # Do not rewrite absolute paths to be relative.
+        no_path_rewrite=1
+
+        # Do not retry
+        retry_limit=0
+
+        # Platform properties
+        REMOTE_PLATFORM_OS=#{OS.mac? ? "macos" : OS.kernel_name.downcase}
+        REMOTE_PLATFORM_ISA=#{Hardware::CPU.arch}
+      EOS
+    end
```
```suggestion
    (buildpath/"recc.conf").write <<~EOS
      ##
      # For the list of configuration parameters please visit:
      #   https://buildgrid.gitlab.io/recc/configuration-variables.html
      #
      # The configuration file settings use lowercase letters without the RECC_ prefix
      # By default recc reads the configuration options from the following places,
      # applying settings bottom-up, with 1 being the last applied configuration.
      # If an option is set in multiple places, the one higher on this list will be the
      # effective one
      #
      # 1. Environment variables
      # 2. ${cwd}/recc/recc.conf
      # 3. {$RECC_CONFIG_DIRECTORY}/recc.conf
      # 4. ~/.recc/recc.conf
      # 5. #{etc}/recc.conf
      ##

      # Protocol version used for communicating with the REAPI server
      reapi_version=2.2

      # The URI of the server
      server=unix://#{var}/recc/casd/casd.sock
      # Instance name provided by the server
      instance=recc-server

      # Use cache-only mode, build anything not available in the cache locally
      cache_only=1
      # Upload results from the local builds to cache
      cache_upload_local_build=1
      # Upload failed results from the local builds to cache
      cache_upload_failed_build=1

      # Enable for link commands
      link=1
      # Use cache-only mode for link commands
      link_cache_only=1

      # Report all entries returned by the dependency command, even if they are absolute paths
      deps_global_paths=1
      # Do not rewrite absolute paths to be relative.
      no_path_rewrite=1

      # Do not retry
      retry_limit=0

      # Platform properties
      REMOTE_PLATFORM_OS=#{OS.mac? ? "macos" : OS.kernel_name.downcase}
      REMOTE_PLATFORM_ISA=#{Hardware::CPU.arch}
    EOS
    etc.install "recc.conf"
```
But to be honest I'm reluctant to have this in the formula -- it seems like something that will easily bitrot here.

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,181 @@
+
+  test do
+    # Start recc server
+    recc_cache_dir=testpath/"recc_cache"
```
```suggestion
    recc_cache_dir = testpath/"recc_cache"
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,181 @@
+    recc_casd_pid = spawn bin/"recc-server", recc_cache_dir
+
+    # Create a source file to test caching
+    test_file=testpath/"test.c"
```
```suggestion
    test_file = testpath/"test.c"
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,181 @@
+    # Override default values of server and log_level
+    ENV["RECC_SERVER"] = "unix://#{recc_cache_dir}/casd.sock"
+    ENV["RECC_LOG_LEVEL"] = "info"
+    recc_test="#{bin}/recc-cc -c #{test_file} 2>&1"
+
+    # Compile the test file twice. The second run should get a cache hit
+    system recc_test
+    output = shell_output(recc_test)
+    assert_match "Action Cache hit", output
```
```suggestion
    recc_test=[bin/"recc-cc", "-c", test_file]

    # Compile the test file twice. The second run should get a cache hit
    system *recc_test
    output = shell_output("#{recc_test.join(" ")} 2>&1")
    assert_match "Action Cache hit", output
```

carlocab commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,181 @@
+  def caveats
+    <<~EOS
+      To launch a compiler with recc, set the following variables:
+        CC=#{bin}/recc-cc
+        CXX=#{bin}/recc-c++
```
```suggestion
        CC=#{opt_bin}/recc-cc
        CXX=#{opt_bin}/recc-c++
```

MikeMcQuaid commented:
@sakeeb Please `git fetch; git rebase -i origin/master` and squash the commits together; right now this PR isn't reviewable unfortunately.

sakeeb commented:
> @sakeeb Please `git fetch; git rebase -i origin/master` and squash the commits together; right now this PR isn't reviewable unfortunately.

Rebased and squashed

MikeMcQuaid commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,180 @@
+
+    # Generate recc.conf
+    (buildpath/"recc.conf").write <<~EOS
+      ##
```
Similarly here.

MikeMcQuaid commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,180 @@
+      (bin/"recc-#{compiler}").write <<~EOS
+        #!/usr/bin/env sh
+
+        #{bin}/recc $(command -v #{compiler}) "$@"
```
Is there any scope to upstream some/all of these helpers scripts?

sakeeb replied:
It makes sense to add the `recc-${compiler}` script. I'll also revisit our previous discussion on defining some default values for those included in `recc.conf`.

If there are no major concerns, would it be acceptable to merge the formula in the meantime?

MikeMcQuaid replied:
> If there are no major concerns, would it be acceptable to merge the formula in the meantime?

I'm afraid this is a fairly major concern to me given how much actual runtime code (not just building the software) is being shipped in this formula. It's the bulk of this formula file and this is unusual in Homebrew.

What's your relationship with upstream? Are you a user/contributor/maintainer?

Would it be tenable to ship a first version of this formulae with no compiler/server helper scripts and an upstream configuration file instead?

sakeeb replied:
I am one of the upstream maintainers.  I would prefer shipping it in out-of-the box working state. I will move the PR to draft state and come back after adding config and helper scripts upstream.

MikeMcQuaid replied:
Thanks! Appreciate your patience and hard work here ❤️

sakeeb replied:
As discussed, I added the wrappers in upstream project and updated the formula to use them

SMillerDev commented:
Dropped the merge commit, this should now be ready for review again

MikeMcQuaid commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,118 @@
+    system "cmake", "--build", "build"
+    system "cmake", "--install", "build"
+
+    ENV["RECC_BIN"] = bin/"recc"
+    system "command envsubst < scripts/wrapper-templates/recc-c++.in > recc-c++"
+    system "command envsubst < scripts/wrapper-templates/recc-cc.in > recc-cc"
+    bin.install "recc-c++"
+    bin.install "recc-cc"
```
A `make install` for this would be nice to have.

MikeMcQuaid commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,118 @@
+    ENV["RECC_REMOTE_PLATFORM_ISA"] = Hardware::CPU.arch.to_s
+    ENV["RECC_REMOTE_PLATFORM_OS"] = OS.kernel_name.downcase
+    ENV["RECC_REMOTE_PLATFORM_OS_VERSION"] = OS.kernel_version.to_s
+    system "command envsubst < scripts/wrapper-templates/recc.conf.in > recc.conf"
+    etc.install "recc.conf"
+
+    bin.install "scripts/wrapper-templates/casd-helper" => "recc-server"
```
And this

MikeMcQuaid commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,118 @@
+  end
+
+  def caveats
+    <<~EOS
+      To launch a compiler with recc, set the following variables:
+        CC=#{opt_bin}/recc-cc
+        CXX=#{opt_bin}/recc-c++
+    EOS
```
Are these Homebrew specific caveats?

MikeMcQuaid commented on Formula/r/recc.rb:
```diff
@@ -0,0 +1,118 @@
+    # Override default values of server and log_level
+    ENV["RECC_SERVER"] = "unix://#{recc_cache_dir}/casd.sock"
+    ENV["RECC_LOG_LEVEL"] = "info"
+    recc_test=[bin/"recc-cc", "-c", test_file]
```
```suggestion
    recc_test = [bin/"recc-cc", "-c", test_file]
```

MikeMcQuaid reviewed (approved):
Thanks @sakeeb, great work (and patience) here. All comments can be addressed in follow-up releases/PRs. You rock!
